Fix selecting weeds which need to be removed

rodja opened the pull request:
This pull request fixes #100 where weeds are skipped. It adds a improved visualization for the camera view which was very helpful for finding the issues in the first place. Note that the visualization will not work as expected with real hardware due to the missing of https://github.com/zauberzeug/rosys/discussions/130.

Also it probably also fixes https://github.com/zauberzeug/field_friend/issues/85.

**ToDos**

- [x] fix punch dialog for weeding screw and tornado
- [x] cleanup logging
- [x] remove boolean return type from `start_workflow`
- [x] test on robot with Weeding Screw (with punch dialog)
- [x] test on robot with Tornado (with punch dialog)

rodja commented:
@pascalzauberzeug I have already found and fixed two issues which resulted in bad (or even no) weeding performance:

- In `weending_implement._has_plants_to_handle` there was a filter which removed weeds behind the robot -- but it used global coordinates instead of local coordinates. Thereby it depended on the robots orientation if the filter worked or not.
- The `detection_time` of a plant was not updated and hence a relevant plant may get pruned by the plant detector.

rodja commented:
After further enhancing the visualization it looks like there are no more weeds skipped:

<img src="https://github.com/zauberzeug/field_friend/assets/131391/2e231d00-b9b4-4431-9880-34ccbf2f64fb" width="300">

rodja commented:
@pascalzauberzeug I think the simulation shows very well that it works. There will be a viszalization error for the plant provider projection because we are still waiting on https://github.com/zauberzeug/rosys/discussions/130.

pascalzauberzeug commented on field_friend/automations/implements/weeding_screw.py:
```diff
@@ -31,13 +32,19 @@ async def start_workflow(self) -> bool:
                 return True
             self.log.info(f'Weeds in range {[f"{p.x:.3f},{p.y:.3f}" for p in weeds_in_range.values()]}')
             for next_weed_id, next_weed_position in weeds_in_range.items():
+                # next_weed_position.x += 0.01  # NOTE somehow this helps to mitigate an offset we experienced in the tests
```
What are we going to do about this?

rodja replied:
I think we should split this into a separate issue where we investigate the offsets between punching and targets.

pascalzauberzeug reviewed (approved):
I did some small refactoring, otherwise it looks good to me.

rodja commented:
Before we merge, we should test it on a real robot -- but the weather is quite bad right now. We hope to test at the end of the week.

Johannes-Thiel commented:
On U6 i have the following problem 
```
| 2024-07-03 13:56:53.634 [ERROR] nicegui/app/app.py:119: unsupported operand type(s) for /: 'NoneType' and 'int'
rosys_1       | Traceback (most recent call last):
rosys_1       |   File "/home/zauberzeug/.local/lib/python3.11/site-packages/nicegui/elements/timer.py", line 94, in _invoke_callback
rosys_1       |     result = self.callback()
rosys_1       |              ^^^^^^^^^^^^^^^
rosys_1       |   File "/app/field_friend/interface/components/camera_card.py", line 141, in update_content
rosys_1       |     svg += self.build_svg_for_implement()
rosys_1       |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
rosys_1       |   File "/app/field_friend/interface/components/camera_card.py", line 212, in build_svg_for_implement
rosys_1       |     tool_2d = self.camera.calibration.project_to_image(tool_3d) / self.shrink_factor
rosys_1       |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~
rosys_1       | TypeError: unsupported operand type(s) for /: 'NoneType' and 'int'
```

rodja commented:
Thanks for pointing this out @Johannes-Thiel. In aab18b7 I hopefully fixed it by not rendering the new augmentation on real robots. As mentioned in https://github.com/zauberzeug/field_friend/pull/102#issuecomment-2200340793, we need to wait for a new RoSys feature.

pascalzauberzeug commented:
We couldn't get this code to work U6/RB34 on a real field.
[#3d5a355](https://github.com/zauberzeug/field_friend/pull/102/commits/3d5a355b2d36f030f620251a2e5af4d82a964c91) adds a test based on real plant positions the robot observed today and [#a06d05e](https://github.com/zauberzeug/field_friend/pull/102/commits/a06d05e90b656268a6637be7cf1d11963aa80a95) fixes the issue.

pascalzauberzeug commented:
All related tests work now, but I think there is much debugging to do, because I saw some weeds skipped in simulation, and some changes feel kind of hacky

rodja commented:
I just came about another problem with the weeding screw. It seems that a bad combination of weeding position and "keep crops safe" will result in endlessly drilling the same spot:

```
2024-07-11 04:53:23.160 [INFO] field_friend/automations/implements/weeding_screw.py:68: Targeting weed 7f987c38-9bea-46cc-a38b-7a5b73463876 which is 0.011487399210734994 away at world: {weed_world_position}, local: {next_weed_position}
2024-07-11 04:53:23.537 [INFO] field_friend/automations/implements/weeding_implement.py:152: Moved weed 7f987c38-9bea-46cc-a38b-7a5b73463876 from Point(0.1, 0.01) to Point(0.1, 0.04) by 0.025218786962445096 to safe 0906bec5-658f-4a43-9599-896c5647cd74 at Point(0.1, 0.0)
2024-07-11 04:53:23.537 [INFO] field_friend/automations/implements/weeding_implement.py:152: Moved weed 720f4c6e-c115-4879-86e7-e8546bc00d4a from Point(0.31, 0.0) to Point(0.33, 0.0) by 0.02505824264348059 to safe 7b5d2efc-2fdc-4b20-9d8d-6729c3c919ae at Point(0.3, -0.0)
2024-07-11 04:53:23.537 [INFO] field_friend/automations/puncher.py:71: Punching at 0.03499512028734873 with depth 0.13...
2024-07-11 04:53:23.537 [INFO] field_friend/hardware/y_axis.py:47: moving yaxis to 0.03499512028734873 with speed 80000
2024-07-11 04:53:24.036 [INFO] field_friend/automations/puncher.py:118: punched at 0.03 with depth 0.13, now back to rest position "reference"
2024-07-11 04:53:24.037 [INFO] field_friend/automations/implements/weeding_screw.py:35: removing weeds at screw world position Point(0.99, 0.03) with radius 0.025
2024-07-11 04:53:24.037 [INFO] field_friend/automations/implements/weeding_implement.py:88: workflow completed
2024-07-11 04:53:24.037 [INFO] field_friend/automations/implements/weeding_implement.py:152: Moved weed 7f987c38-9bea-46cc-a38b-7a5b73463876 from Point(0.1, 0.01) to Point(0.1, 0.04) by 0.02555860171762547 to safe 0906bec5-658f-4a43-9599-896c5647cd74 at Point(0.1, 0.0)
2024-07-11 04:53:24.038 [INFO] field_friend/automations/implements/weeding_implement.py:152: Moved weed 720f4c6e-c115-4879-86e7-e8546bc00d4a from Point(0.31, -0.0) to Point(0.33, -0.0) by 0.02483071511458918 to safe 7b5d2efc-2fdc-4b20-9d8d-6729c3c919ae at Point(0.3, -0.0)
2024-07-11 04:53:24.038 [INFO] field_friend/automations/implements/weeding_screw.py:51: Found 9 weeds in range: {'7f987c38-9bea-46cc-a38b-7a5b73463876': Point(x=0.09669951415087447, y=0.035313021484457204), '67634fe7-8053-4307-9992-a3ed4857752b': Point(x=0.11704021151318222, y=-0.029777316000581693), 'd181ab5c-37d3-4541-a7ae-7545c1bd349f': Point(x=0.13661206769921896, y=-0.05026863826333809), '2690d481-cce8-496f-8e3e-debf0e49ae25': Point(x=0.1467635056140576, y=-0.08017540718766214), '9d43317a-f4ba-412b-bc93-25d57a3727c6': Point(x=0.24661214437973544, y=-0.030140018155814973), '8e45ffcb-4cc5-4b45-b3e2-8aa3828dd565': Point(x=0.2666659087491333, y=-0.059915353889255396), '5bd33a65-c626-4dba-893d-c9ba05a37026': Point(x=0.3265859465999482, y=0.040039277022019024), '720f4c6e-c115-4879-86e7-e8546bc00d4a': Point(x=0.3316018667031977, y=-0.0003072817664132812), '3a4bf335-17a8-440c-bb4a-a10a74a1251f': Point(x=0.336689218715155, y=-0.07994842985984071)}
2024-07-11 04:53:24.038 [INFO] field_friend/automations/implements/weeding_screw.py:68: Targeting weed 7f987c38-9bea-46cc-a38b-7a5b73463876 which is 0.01169951415087446 away at world: {weed_world_position}, local: {next_weed_position}
2024-07-11 04:53:24.415 [INFO] field_friend/automations/implements/weeding_implement.py:152: Moved weed 7f987c38-9bea-46cc-a38b-7a5b73463876 from Point(0.1, 0.01) to Point(0.1, 0.04) by 0.02547690550984423 to safe 0906bec5-658f-4a43-9599-896c5647cd74 at Point(0.1, 0.0)
2024-07-11 04:53:24.415 [INFO] field_friend/automations/implements/weeding_implement.py:152: Moved weed 720f4c6e-c115-4879-86e7-e8546bc00d4a from Point(0.31, -0.0) to Point(0.33, -0.0) by 0.02499742286996728 to safe 7b5d2efc-2fdc-4b20-9d8d-6729c3c919ae at Point(0.3, 0.0)
2024-07-11 04:53:24.415 [INFO] field_friend/automations/puncher.py:71: Punching at 0.035313021484457204 with depth 0.13...
2024-07-11 04:53:24.415 [INFO] field_friend/hardware/y_axis.py:47: moving yaxis to 0.035313021484457204 with speed 80000
2024-07-11 04:53:24.912 [INFO] field_friend/automations/puncher.py:118: punched at 0.04 with depth 0.13, now back to rest position "reference"
2024-07-11 04:53:24.913 [INFO] field_friend/automations/implements/weeding_screw.py:35: removing weeds at screw world position Point(0.99, 0.04) with radius 0.025
2024-07-11 04:53:24.913 [INFO] field_friend/automations/implements/weeding_implement.py:88: workflow completed
2024-07-11 04:53:24.913 [INFO] field_friend/automations/implements/weeding_implement.py:152: Moved weed 7f987c38-9bea-46cc-a38b-7a5b73463876 from Point(0.1, 0.01) to Point(0.1, 0.03) by 0.025111894078605347 to safe 0906bec5-658f-4a43-9599-896c5647cd74 at Point(0.1, -0.0)
2024-07-11 04:53:24.914 [INFO] field_friend/automations/implements/weeding_implement.py:152: Moved weed 720f4c6e-c115-4879-86e7-e8546bc00d4a from Point(0.31, -0.0) to Point(0.33, -0.0) by 0.024799826714327303 to safe 7b5d2efc-2fdc-4b20-9d8d-6729c3c919ae at Point(0.3, 0.0)
2024-07-11 04:53:24.914 [INFO] field_friend/automations/implements/weeding_screw.py:51: Found 9 weeds in range: {'7f987c38-9bea-46cc-a38b-7a5b73463876': Point(x=0.0956527758332895, y=0.03486667650294348), '67634fe7-8053-4307-9992-a3ed4857752b': Point(x=0.1166819950373813, y=-0.029893607904045816), 'd181ab5c-37d3-4541-a7ae-7545c1bd349f': Point(x=0.13665047922347984, y=-0.05003241327292461), '2690d481-cce8-496f-8e3e-debf0e49ae25': Point(x=0.14682159508869874, y=-0.07999531694631856), '9d43317a-f4ba-412b-bc93-25d57a3727c6': Point(x=0.24662827175536883, y=-0.029764111672987294), '8e45ffcb-4cc5-4b45-b3e2-8aa3828dd565': Point(x=0.26656134654734254, y=-0.05993003899133466), '5bd33a65-c626-4dba-893d-c9ba05a37026': Point(x=0.32668532160176234, y=0.039951812953086745), '720f4c6e-c115-4879-86e7-e8546bc00d4a': Point(x=0.33157048127307714, y=-0.00037669166490071594), '3a4bf335-17a8-440c-bb4a-a10a74a1251f': Point(x=0.3366683598025787, y=-0.07997475902251638)}
2024-07-11 04:53:24.914 [INFO] field_friend/automations/implements/weeding_screw.py:68: Targeting weed 7f987c38-9bea-46cc-a38b-7a5b73463876 which is 0.010652775833289496 away at world: {weed_world_position}, local: {next_weed_position}
```

rodja commented:
I could not find a good reproduction in an automated test. But in 170e774 we used the safety distance (0.01) instead of drill radius (0.025) which explains why we kept drilling instead of skipping the position after one attempt.

rodja commented:
The live test with U6 worked flawlessly.

pascalzauberzeug commented:
I just tested it on F13 with a tornado, works!

pascalzauberzeug commented:
Works on U5 aswell, I will merge now.
